### ISSUE TITLE
Possible variable truncation on 32bit platform

### ISSUE THREAD

**Issue by HolakovskyJan** [civetweb/civetweb#488]
In function `read_websocket` on line 10649:
`data_len = (((uint64_t)ntohl(l1)) << 32) + ntohl(l2);`
The cast to `uint64_t` and shift will do nothing on 32bit platform because `data_len` is `size_t` (only 32bit on 32bit platform).

**Comment by HolakovskyJan** [comment]
And similar problem is in function `mg_websocket_write_exec` on line 10810:
`uint32_t len1 = htonl((uint32_t)((uint64_t)dataLen >> 32));`

**Comment by bel2125** [comment]
The 32 bit platform does not support > 32 bit data size blocks anyway.

In 10649 we have
`size_t data_len = (((uint64_t)ntohl(l1)) << 32) + ntohl(l2);`

and in line 10657:
`unsigned char *data = (unsigned char *)mg_malloc_ctx(data_len, conn->ctx);`

But we cannot allocate more than 2 GB on a 32 bit platform anyway.
So it does not really help us to use int64_t instead of size_t ***here*** ... although I come more and more to the conclusion, we should get rid of size_t completely, and replace it by int64_t.


I just wonder why I did not get a warning

**Comment by bel2125** [comment]
I think for the moment we can just limit the websocket payload to 2 GB.
I've never ever seen a websocket package even close to that size.

Probably one could do it only `if (sizeof(size_t) < 64)`, but that causes "condition is always true/false" warnings, and I think there is no way to check this in a preprocessor `#if`

**Comment by manison** [comment]
> But we cannot allocate more than 2 GB on a 32 bit platform anyway.

Although partially true it does not mean you cannot handle 64b numbers correctly on 32b platform. Imagine the websocket frame is indicated to be 0x1234_5678_0000_0005 bytes long. The code above will truncate (provided `size_t` is 32b) the value to 0x0000_0005 bytes by throwing the upper 32 bits away. And surely you can `malloc` that amount of bytes even on 32b platform. The behavior in this case will be wrong.

I suggest handling websocket frame sizes explicitly in `(u)int64_t` and introducing some configuration option to limit the maximum frame size (also for memory consumption reasons). This limit could be then set as high as 2^32 on 32b platforms and 2^63 on 64b platforms.

**Comment by bel2125** [comment]
@manison: I've been on vacation, and did not check mails - sorry for the late reply.

I've introduced a check already a week ago:
https://github.com/civetweb/civetweb/commit/494c957eddf137256afa71aa37d17c2443bbb7d1

It will no longer truncate websocket messages, but raise an error and close the connection.
I've never seen websocket messages with more than a few kB in practice - upload and download of files, of course, maybe a but never a single websocket message.
The current API for websocket messages is a callback, with a buffer of data + length of the message. If huge websocket message should be stored, we would need a different (additional) API, like `handle_form`. 
I would not add this new API, if there is no real reason to do so. 
Yes, in theory one could send 100 GB in a single websocket message, but we should not add and maintain an API will never be used in practice.

So I think, this check should be sufficient, unless there is a use case I don't s

**Comment by bel2125** [comment]
The last paragraph in the last post is somehow ... incomplete, and I can't remember what I intended to tell. Anyway, there is now a limit

see here: https://github.com/civetweb/civetweb/blob/46dc39cc4fbc8f9e5b1a673dc416bdd7ad5f8106/src/civetweb.c#L11172

A 32 bit system will run out of memory before reaching this limit. A 64 bit system will not, but I found even on 64 bit systems some API functions have 31/32 bit size limits (e.g., https://github.com/civetweb/civetweb/blob/46dc39cc4fbc8f9e5b1a673dc416bdd7ad5f8106/src/civetweb.c#L8612).

For the moment, 2^31 (- 1 page) appears sufficient as the maximum size of a single websocket message.
